### PR TITLE
[WFLY-18163] Make it possible to use JaasSecurityRealm via a custom-realm resource

### PR DESCRIPTION
**Pull request by Skyllarr**
https://issues.redhat.com/browse/WFLY-18163

### CODE REVIEW

File: elytron/ELY-2544-jaas-realm-via-custom-realm.adoc [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+== Implementation Plan
+
+The task will be achieved by adding custom realm wrapper around existing JaasSecurityRealm class. This custom realm wrapper will be deprecated as users should use jaas realm directly when available. This wrapper will be added to the `wildfly-core` repository and can be utilized by accessing a module `modules/system/layers/base/org/wildfly/extension/elytron`.
+
+Example of usage:
+
+```
+[standalone@localhost:9990 /] module add --name=jaasLoginModule --resources=path/to/login/modules/jar --dependencies=org.wildfly.security.elytron
+[standalone@localhost:9990 /] module add --name=elytronExtensionModule --dependencies=org.wildfly.extension.elytron,org.wildfly.security.elytron --resources=modules/system/layers/base/org/wildfly/extension/elytron
+[standalone@localhost:9990 /] /subsystem=elytron/custom-realm=cd:add(module=elytronExtensionModule,class-name=org.wildfly.extension.elytron.JaasCustomSecurityRealmWrapper,configuration={entry=Entry1[standalone@localhost:9990 /] module=jaasLoginModule,callback-handler=org.example.ExampleCallbackHandler,path=/my/example/path,relative-to=jboss.server.config.dir})
```
Comment (OndrejKotek) [inline elytron/ELY-2544-jaas-realm-via-custom-realm.adoc:69]: > `{entry=Entry1[standalone@localhost:9990 /] module=`

-> `{entry=Entry1,module=` ?

File: elytron/ELY-2544-jaas-realm-via-custom-realm.adoc [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+* WildFly Core test suite: Functional tests - testing of adding custom-realm resource that uses JaasSecurityRealm class
+* WildFly test suite: arquillian tests using a deployed web app secured with jaas realm defined as a custom realm resource
+
+== Community Documentation
+
+No community documentation is needed as users can use jaas realm directly.
+
+== Release Note Content
+
+TODO
```
Comment (OndrejKotek) [inline elytron/ELY-2544-jaas-realm-via-custom-realm.adoc:84]: Still TODO

File: elytron/ELY-2544-jaas-realm-via-custom-realm.adoc [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+== Implementation Plan
+
+The task will be achieved by adding custom realm wrapper around existing JaasSecurityRealm class. This custom realm wrapper will be deprecated as users should use jaas realm directly when available. This wrapper will be added to the `wildfly-core` repository and can be utilized by accessing a module `modules/system/layers/base/org/wildfly/extension/elytron`.
+
+Example of usage:
+
+```
+[standalone@localhost:9990 /] module add --name=jaasLoginModule --resources=path/to/login/modules/jar --dependencies=org.wildfly.security.elytron
+[standalone@localhost:9990 /] module add --name=elytronExtensionModule --dependencies=org.wildfly.extension.elytron,org.wildfly.security.elytron --resources=modules/system/layers/base/org/wildfly/extension/elytron
```
Comment (OndrejKotek) [inline elytron/ELY-2544-jaas-realm-via-custom-realm.adoc:68]: Why is this needed? Why do we let this step up to users?

File: elytron/ELY-2544-jaas-realm-via-custom-realm.adoc [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+* https://github.com/wildfly-security/wildfly-elytron[WildFly Elytron project]
+* https://github.com/wildfly-core/wildfly-core will have tests of custom-realm resource added
+* https://github.com/wildfly/wildfly will have community documentation added
+
+== Requirements
+
+=== Hard Requirements
+
+* It must be possible to add the `custom-realm` resource in the management model that will use the code of jaas-realm without an admin having to add any new classes as a module to the WildFly server.
```
Comment (OndrejKotek) [inline elytron/ELY-2544-jaas-realm-via-custom-realm.adoc:50]: > use the code of jaas-realm without an admin having to add any new classes as a module to the WildFly server

Do I understand well that by "new classes" we mean the wrapper and not mean the user defined login module?

File: elytron/ELY-2544-jaas-realm-via-custom-realm.adoc (hunk 3 above)
Comment (Skyllarr) [reply elytron/ELY-2544-jaas-realm-via-custom-realm.adoc:68]: @OndrejKotek Thanks for review! I will update the PR soon. This module is needed for exposing the `org.wildfly.extension.elytron.JaasCustomSecurityRealmWrapper` class so it can be referenced in the subsystem configuration

File: elytron/ELY-2544-jaas-realm-via-custom-realm.adoc (hunk 1 above)
Comment (Skyllarr) [reply elytron/ELY-2544-jaas-realm-via-custom-realm.adoc:69]: Sry this should be: 

[standalone@localhost:9990 /] /subsystem=elytron/custom-realm=myRealm:add(module=elytronExtensionModule,class-name=org.wildfly.extension.elytron.JaasCustomSecurityRealmWrapper,configuration={entry=Entry1,module=jaasLoginModule,callback-handler=org.example.ExampleCallbackHandler,path=/my/example/path,relative-to=jboss.server.config.dir})

File: elytron/ELY-2544-jaas-realm-via-custom-realm.adoc (hunk 4 above)
Comment (Skyllarr) [reply elytron/ELY-2544-jaas-realm-via-custom-realm.adoc:50]: Yes that's right. I will rephrase this to something like:

It must be possible to add the `custom-realm` resource in the management model that will use the code of jaas security realm without an admin having to implement a jaas realm functionality themselves.

File: elytron/ELY-2544-jaas-realm-via-custom-realm.adoc (hunk 3 above)
Comment (Skyllarr) [reply elytron/ELY-2544-jaas-realm-via-custom-realm.adoc:68]: This is because some class name has to be defined for the custom realm. I was thinking previously about adding a configuration parameter that would specify to use the wrapper like so:

```
/subsystem=elytron/custom-realm=myRealm:add(module=?,class-name=?,configuration={jaas-realm=true,entry=Entry1,module=jaasLoginModule,callback-handler=org.example.ExampleCallbackHandler,path=/my/example/path,relative-to=jboss.server.config.dir})
```

but then what would be defined under the module and class-name

**Comment by yersan** [comment]
Just a minor note, I understand this PR will need to add the required transformers and bump the Elytron model.
And one suggestion, if it is in progress, you can use a draft to express the current PR status/intention, Jobs are also executed on Draft PRs.

**Review by fjuma — APPROVED** [review]

File: elytron/ELY-2544-jaas-realm-via-custom-realm.adoc [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+* https://issues.redhat.com/browse/EAP7-2068[EAP7-2068] - EAP issue tracker
+* https://issues.redhat.com/browse/WFCORE-6411[WFCORE-6411] - Wildfly core custom-realm tests issue tracker
+
+=== Dev Contacts
+
+* mailto:dvilkola@redhat.com[Diana Krepinska]
+
+=== QE Contacts
+
+* mailto:TODO
```
Comment (darranl) [inline elytron/ELY-2544-jaas-realm-via-custom-realm.adoc:31]: I think you can add this one now?

File: elytron/ELY-2544-jaas-realm-via-custom-realm.adoc [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+* WildFly Core test suite: Functional tests - testing of adding custom-realm resource that uses JaasSecurityRealm class
+* WildFly test suite: arquillian tests using a deployed web app secured with jaas realm defined as a custom realm resource
+
+== Community Documentation
+
+No community documentation is needed as users can use jaas realm directly.
+
+== Release Note Content
+
+You can now add a JAAS security realm as a custom realm resource without using jaas-realm. This approach is deprecated, and you should use the jaas-realm resource instead.
```
[the remark this answers is not in the document]
Comment (darranl) [reply elytron/ELY-2544-jaas-realm-via-custom-realm.adoc:84]: +1 we can probably omit this one from the release notes.

File: elytron/ELY-2544-jaas-realm-via-custom-realm.adoc [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+== Implementation Plan
+
+The task will be achieved by adding custom realm wrapper around existing JaasSecurityRealm class. This custom realm wrapper will be deprecated as users should use jaas realm directly when available. This wrapper class will be added to the `wildfly-core` repository and new module `org.wildfly.extension.elytron.jaas-realm` will be added to expose it.
+
+Example of usage:
+
+```
+[standalone@localhost:9990 /] module add --name=jaasLoginModule --resources=path/to/login/modules/jar --dependencies=org.wildfly.security.elytron
+
+[standalone@localhost:9990 /] /subsystem=elytron/custom-realm=myRealm:add(module=elytronExtensionModule,class-name=org.wildfly.extension.elytron.JaasCustomSecurityRealmWrapper,configuration={entry=Entry1,module=jaasLoginModule,callback-handler=org.example.ExampleCallbackHandler,path=/my/example/path,relative-to=jboss.server.config.dir})
```
Comment (darranl) [inline elytron/ELY-2544-jaas-realm-via-custom-realm.adoc:69]: Just to be complete it may be good to list the set of supported configuration options so we have a clear list in the analysis.

**Review by darranl — COMMENTED** [review]
Only minor comments but otherwise looks good, most importantly this proposal is just adding the support for a custom component - the JAAS Realm itself is previously defined.

File: elytron/ELY-2544-jaas-realm-via-custom-realm.adoc (hunk 5 above)
Comment (Skyllarr) [reply elytron/ELY-2544-jaas-realm-via-custom-realm.adoc]: Updated

File: elytron/ELY-2544-jaas-realm-via-custom-realm.adoc (hunk 7 above)
Comment (Skyllarr) [reply elytron/ELY-2544-jaas-realm-via-custom-realm.adoc:69]: Updated

**Review by OndrejKotek — APPROVED** [review]

File: elytron/ELY-2544-jaas-realm-via-custom-realm.adoc [hunk 8]
```diff
@@ -0,0 +1,100 @@
+= WFLY-18163 Make it possible to use JaasSecurityRealm via a custom-realm resource
```
Comment (bstansberry) [inline elytron/ELY-2544-jaas-realm-via-custom-realm.adoc:1]: @Skyllarr Please add jekyll front matter above this line so it shows up in the appropriate categories on https://docs.wildfly.org/wildfly-proposals/

Categories are shown in https://github.com/wildfly/wildfly-proposals/blob/main/_data/wildfly-categories.yaml

The first four lines in https://raw.githubusercontent.com/wildfly/wildfly-proposals/main/security/WFCORE-3873_SNI_Support.adoc are an example of what I mean. Almost all currently merged proposals are also examples.

Please ping me in zulip when done and I'll merge this, as the feature itself is merged. Woohoo!

Comment (Skyllarr) [reply elytron/ELY-2544-jaas-realm-via-custom-realm.adoc:5]: @bstansberry I added the categories on top, thank you!

**Comment by bstansberry** [comment]
Thanks @Skyllarr, @fjuma and @OndrejKotek!
